client/src/controllers/SwapController.ts

```diff
@@ -0,0 +1,239 @@
+    const { signal } = this.abortController;
+
+    /** Parse a request URL from the supplied param, as a string or inside a custom event */
+    const requestUrl =
+      typeof param === 'string' ? param : param.detail.url || '';
+
+    this.loadingValue = true;
+
+    this.dispatch('start', {
+      cancelable: false,
+      detail: { requestUrl },
+      // Stimulus dispatch target element type issue https://github.com/hotwired/stimulus/issues/642
+      target: this.targetElement as HTMLInputElement,
+    }) as CustomEvent<{ requestUrl: string }>;
+
+    return fetch(requestUrl, {
+      headers: { 'x-requested-with': 'XMLHttpRequest' },
+      signal,
+    })
+      .then((response) => {
+        if (!response.ok) {
+          throw new Error(`HTTP error! Status: ${response.status}`);
+        }
+        return response.text();
+      })
+      .then((results) => {
+        const targetElement = this.targetElement as HTMLElement;
+        targetElement.innerHTML = results;
+
+        this.dispatch('success', {
+          cancelable: false,
+          detail: { results },
+          // Stimulus dispatch target element type issue https://github.com/hotwired/stimulus/issues/642
+          target: targetElement as HTMLInputElement,
+        });
+      })
+      .catch((error) => {
+        if (error.name === 'AbortError') return;
+        this.dispatch('error', {
+          cancelable: false,
+          detail: { error },
+          // Stimulus dispatch target element type issue https://github.com/hotwired/stimulus/issues/642
+          target: this.targetElement as HTMLInputElement,
+        });
+        // eslint-disable-next-line no-console
+        console.error(`Error fetching ${requestUrl}`, error);
+        delete this.abortController;
+      })
+      .finally(() => {
+        this.loadingValue = false;
+      });
+  }
+
+  /**
+   * Perform a search based on a single input query, and only if that query's value
+   * differs from the current matching URL param. Once complete, update the URL param.
+   * Additionally, clear the 'page' param in the URL if present, can be overridden
+   * via action params if needed.
+   */
+  swapLocation({ params = {} }: { params?: { clear?: string } }) {
+    const clearParams = (params?.clear || 'p').split(' ');
+    const searchInput = this.hasInputTarget ? this.inputTarget : this.element;
+    const queryParam = searchInput.name;
+    const searchParams = new URLSearchParams(window.location.search);
+    const currentQuery = searchParams.get(queryParam) || '';
+    const newQuery = searchInput.value || '';
+
+    // only do the query if it has changed for trimmed queries
+    // for example - " " === "" and "first word " ==== "first word"
+    if (currentQuery.trim() === newQuery.trim()) return;
+
+    // Update search query param ('q') to the new value or remove if empty
+    if (newQuery) {
+      searchParams.set(queryParam, newQuery);
+    } else {
+      searchParams.delete(queryParam);
```

Comment:
I might've missed it but it seems we don't have a test case that clears the input after it was previously entered

Reply:
Good call, have added now.